tensorflow/core/lib/random/random_distributions.h

```diff
@@ -23,22 +23,62 @@ limitations under the License.
 #include <algorithm>
 #include <type_traits>
 
-#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
+#include "tensorflow/core/framework/tensor.h"
 #include "tensorflow/core/lib/bfloat16/bfloat16.h"
 #include "tensorflow/core/lib/random/philox_random.h"
+#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
 
 namespace tensorflow {
 namespace random {
 
 // Helper function to convert a 16-bit integer to a half between [0..1).
 PHILOX_DEVICE_INLINE Eigen::half Uint16ToHalf(uint16 x);
+// Helper function to convert a 16-bit integer to a bfloat16 between [1..2).
+PHILOX_DEVICE_INLINE bfloat16 InternalUint16ToBfloat16(uint16 x);
 // Helper function to convert a 16-bit integer to a bfloat16 between [0..1).
-PHILOX_DEVICE_INLINE bfloat16 Uint16ToGfloat16(uint16 x);
+PHILOX_DEVICE_INLINE bfloat16 Uint16ToBfloat16(uint16 x);
+// Helper function to convert a 32-bit integer to a float between [1..2).
+PHILOX_DEVICE_INLINE float InternalUint32ToFloat(uint32 x);
 // Helper function to convert a 32-bit integer to a float between [0..1).
 PHILOX_DEVICE_INLINE float Uint32ToFloat(uint32 x);
 // Helper function to convert two 32-bit integers to a double between [0..1).
 PHILOX_DEVICE_INLINE double Uint64ToDouble(uint32 x0, uint32 x1);
 
+// Helper function to format distribution result in vectorization path,
+// it creates Eigen::Tensor and reuses packet feature with SIMD.
+template <typename Distribution, typename Generator>
+typename Distribution::ResultType VectorizedFormat(
+    Generator* gen, typename Distribution::FormatFunc functor) {
+  typename Generator::ResultType sample;
+  typename Distribution::ResultType result;
+  const int kResultElementCount = Distribution::kResultElementCount;
+  const int inner_count = Generator::kResultElementCount;
+  const int outer_count = kResultElementCount / inner_count;
+  int offset = 0;
+
+  for (int k = 0; k < outer_count; k++) {
+    sample = (*gen)();
+    for (int i = 0; i < inner_count; i++, offset++) {
+      result[offset] = (*functor)(sample[i]);
+    }
+  }
+  // Tail processing if any.
+  // Put the tail condition out of above loop to improve performance:
+  // it will be executed only once and save time on CPU.
+  if (offset < kResultElementCount) {
+    sample = (*gen)();
+    for (int i = 0; offset < kResultElementCount; i++, offset++) {
+      result[offset] = (*functor)(sample[i]);
+    }
+  }
+
+  auto tensor_result =
+      typename TTypes<typename Distribution::ResultElementType>::Tensor(
+          &result[0], kResultElementCount);
```

Comment:
I have tried this solution locally, too. But I got another failure `//tensorflow/core:lib_random_random_distributions_test`:
```
In file included from external/eigen_archive/unsupported/Eigen/CXX11/../../../Eigen/Core:153:0,
                 from external/eigen_archive/unsupported/Eigen/CXX11/Tensor:14,
                 from ./third_party/eigen3/unsupported/Eigen/CXX11/Tensor:1,
                 from ./tensorflow/core/lib/random/random_distributions.h:28,
                 from tensorflow/core/lib/random/random_distributions_test.cc:16:
external/eigen_archive/unsupported/Eigen/CXX11/src/Tensor/TensorBase.h:
...
tensorflow/core/lib/random/random_distributions_test.cc:261:79:   required from here
external/eigen_archive/unsupported/Eigen/CXX11/src/Tensor/TensorBase.h:354:7: error: static assertion failed: YOU_MADE_A_PROGRAMMING_MISTAKE
       EIGEN_STATIC_ASSERT((NumTraits<Scalar>::IsSigned || internal::is_same<Scalar, const std::complex<float> >::value), YOU_MADE_A_PROGRAMMING_MISTAKE);
```

Although TF has defined `NumTraits<Scalar>::IsSigned` for bfloat16 [here](https://github.com/Intel-tensorflow/tensorflow/blob/tenglu/vectorize_random_uniform/tensorflow/core/framework/numeric_types.h#L67), seems this case can't find it with direct Eigen TensorMap API. Howerver, it can works well after I move bfloat16 definition to Eigen.

Reply:
So currently I get 2 optional:
1. For some reasons Eigen TensorMap implementation got dependency on Eigen BF16 PR. I can't use this implementation until Eigen BF16 PR is merged or find other way to resolve UT failures.
2. Modify the BUILD dependency from "//tensorflow/core:framework" to "//tensorflow/core/framework:kernel_shape_util". Because the original dependency cycle is generated from:
```
//tensorflow/core:framework
└ //tensorflow/core:lib
    └ //tensorflow/core/lib/random:philox
        └ //tensorflow/core:framework
```
After this modification it becomes like:
```
//tensorflow/core:framework
├ //tensorflow/core/framework:kernel_shape_util
└ //tensorflow/core:lib
    └ //tensorflow/core/lib/random:philox
        └ //tensorflow/core/framework:kernel_shape_util
```

I'll keep try to resolve option 1 first, but please let me know if you think option 2 is acceptable, thanks.

Reply:
`NumTraits` is defined in `tensorflow/core/framework/numeric_types.h` and it's not included here. (It's included in `tensorflow/core/framework/bfloat16.h`, not the `tensorflow/core/lib/bfloat16/bfloat16.h` we use. I believe including `numeric_types.h` and its target (`//tensorflow/core/framework:numeric_types`) should solve the problem. 

Let's try this first and see if anything else fails?